### PR TITLE
feat(irc): resilient connection & channel handling with improved auto-recover

### LINKED ISSUE #1529: Memory leak in Settings > IRC > View
**Version**
dev

**Describe the bug**
Memory usage of tab explodes at some point

**To Reproduce**
1. Go to 'Settings'
2. Click on 'IRC'
3. Select an active channel
4. Click on View
5. Wait and monitor memory usage of tab

**Expected behavior**
No memory leak

**Desktop (please complete the following information):**
 - OS: Arch
 - Browser Firefox
 - Version 125

**Additional context**
I have the impression that the current behaviour doesn't add new lines that were received, but adds the most recent $number of lines. The number appears to be large and thus memory usage of the tab starts increasing and increasing with each new announcement. You can also find duplicate announcements by scrolling and searching through the growing list of announces.

### PR DESCRIPTION
**Pull request by zze0s**
#### What is the relevant ticket/issue

* Fixes #1529 

Refactors the IRC implementation (`internal/irc/`) into explicit connection and channel state machines. The previous handler interleaved connection lifecycle, authentication, channel joining, and announce processing with ad-hoc flags and goroutines, which made reconnects, invite-only channels, kicks, key-protected channels, and auth timing fragile, and left shared state with data races.

##### Add

* **Two explicit state machines**:
  * Connection (`state_machine.go`): `Disconnected -> Connecting -> Connected -> Authenticating -> Authenticated -> JoiningChannels -> FullyOperational` (or `PartiallyOperational` / `Error`). Owns auth (None / SASL / NickServ) and drives joins.
  * Per-channel (`channel_state_machine.go`): `Idle -> AwaitingInvite / AwaitingInviteBot / InviteFailed(NoSuchNick) -> Joining -> Monitoring`, plus `Kicked / Parted / Disabled / Error`, with bounded exponential backoff and generation-guarded timeouts so a stalled invite/join can't hang forever.
* **Specific join-failure handling** - numerics `471`/`473`/`474`/`475`/`477` (full / invite-only / banned / bad-key / needs-registered-nick) drive the affected channel to `Error` with a human-readable reason instead of a 45s stall.
* **Invite flow** - invite command -> `INVITE` -> join. A present bot that answers with a rejection parks the channel in `InviteFailed` with the reason surfaced and stops (no re-spamming a bot that refused us); an absent bot (`401` / silent timeout) keeps retrying via backoff; a parked channel recovers on a later `INVITE`, config change, or reconnect.
* **On-the-fly channel reconcile** - adding, removing, enabling/disabling, or re-keying a channel is applied live (`JOIN`/`PART`/rejoin) without restarting the whole network connection.
* **UI** - per-channel enable toggle in the network form; channel state pills show the specific error reason; new `useIrcEvents` hook consolidating SSE handling with a polling fallback.
* **In-process IRC integration test suite** (`test/irc`, behind the `irc_integration_test` build tag, no Docker/binaries): a minimal Go IRC server (CAP + SASL PLAIN, NickServ, channel modes `+k`/`+i`/`+r`, scriptable announcer/invite-gatekeeper bots) plus a harness running the real `Handler` over a loopback socket, including a table-driven test across all 16 tracker configs. Plus unit tests for reconnect, recovery/backoff, join errors, health, and invite park/retry.

##### Change

* **Reconnect trap fixed** - channels reset to `Idle` on disconnect so they rejoin after ircevent's auto-reconnect. Previously `Monitoring` had no path back, so a dropped connection left channels permanently stuck and announces silently stopped until a process restart.
* **Channel keys (`+k`)** are now sent on `JOIN`.
* **Kicked channels** park in `Kicked` and are not auto-rejoined (deliberate, to avoid bans); they clear only on reconnect, restart, or an explicit invite.
* **Network health** is driven only by enabled announce channels, so a failing user-added extra channel no longer marks the whole network unhealthy.
* **Concurrency** - `Channel`/`Handler` shared state is guarded with consistent lock discipline; several data races fixed (including one in the connection state machine caught by the new integration tests under `-race`).

##### Remove

* The previous Docker + Ergo-config integration setup (`test/integration/ircd.yaml`, `docker-compose.yml`), replaced by the in-process `test/irc` harness.

#### Where should the reviewer start?

* `internal/irc/state_machine.go` and `internal/irc/channel_state_machine.go` - the state tables and `transition()`. Note that ground-truth events (kick/part/error/reset) set state directly, everything else goes through a validated transition.
* `internal/irc/handler.go` - how the IRC callbacks map onto the state machines.
* `test/irc/` - the integration harness and scenarios (a good end-to-end read of the intended behavior).

#### How should this be manually tested?

* Connect networks using each auth type (None / SASL / NickServ); join a direct channel and an invite-only channel; add a `+k` channel with and without the key; get kicked; drop the connection and confirm channels rejoin. Confirm the per-channel toggle and that specific errors (e.g. `475`) show on the channel status.
* Automated: `go test -tags=irc_integration_test ./test/irc/...` (add `-race`). CI runs these on the ubuntu leg only.

#### Risk involved?

* Core IRC path is rewritten, so the main risk is connection-lifecycle / reconnect edge cases. Mitigated by making the lifecycle explicit and by broad unit + integration coverage (race-clean). No database migration. No new runtime dependencies - the test server is build-tag gated and never ships in the binary.

#### Screenshots (if appropriate)

* Screenshot_20260705_005428
* image
* Screenshot_20260705_005302

#### Does the documentation or dependencies need an update?

* No new runtime dependencies. Test-only tooling (`test/irc`) is gated behind the `irc_integration_test` build tag and is not part of the shipped binary.

### CODE REVIEW

File: internal/irc/service.go [hunk 1]
```diff
@@ -68,7 +72,8 @@ func NewService(log logger.Logger, sse *sse.Server, repo domain.IrcRepo, release
 		indexerService:      indexerSvc,
 		notificationService: notificationSvc,
 		proxyService:        proxySvc,
-		handlers:            make(map[int64]*Handler),
+		networkCache:        ttlcache.New[int64, *domain.IrcNetwork](ttlcache.WithTTL[int64, *domain.IrcNetwork](5 * time.Minute)),
```
[the remark this answers is not in the document]
Comment (zze0s) [reply internal/irc/service.go:75]: The idea was to only cache this if a user visits the irc settings and then clear it until the next time, which could be a month later.

There was however a few annoyances with this optimization so I dropped the cache usage in the `GetNetworksWithHealth` method. The webui refreshes the networks every 3 seconds and the original implementation is good enough for this.

File: internal/domain/irc.go [hunk 2]
```diff
@@ -92,13 +93,26 @@ type ChannelWithHealth struct {
...
+}
+
+func (u *IrcUser) ParseMode(nick string) bool {
+	index := strings.IndexAny(nick, "~!@+&")
+	if index == -1 {
+		return false
+	}
+
+	u.Mode = nick[:index+1]
+	u.Nick = nick[index+1:]
```
[the remark this answers is not in the document]
Comment (zze0s) [reply internal/domain/irc.go:113]: It does a check for the characters first `index := strings.IndexAny(nick, "~!@+&")` and if it's not found it returns early, so it should have the contents if it reaches those lines unless I'm missing something.

I don't think those are valid characters in an irc nick and only used by the server to set modes etc?

**Comment by zze0s** [comment]
Did some more testing on this and it mostly works but have issues when the perms of the channel are too strict and the announcer isn't shown as being part of the channel. It reports the server as unhealthy because of this and it seems to not process the announces or at least not update `Last announce`. Will run with debugger and see what actually happens.
